`bitsandbytes` - `Linear8bitLt` integration into `transformers` models

younesbelkada opened the pull request:
# What does this PR do?

Adding the `bitsandbytes` - `Linear8bitLt` integration for large language models! 🚀 
This feature could reduce the size of the large models by up to 2, without a high loss in precision
Paper and main implementations from: @TimDettmers

# Usage:

Anyone with a GPU that supports mixed 8 bit quantization could load a model using
`AutoModel.from_pretrained(xxx, load_in_8bit=True, device_map="auto")` 
And works like charm. Could work on *any* HF model!

## Requirements

Needs the latest version of `bitsandbytes` (that is compiled manually) and `accelerate` 

## TODOs:

- [x] Add custom tests
- [x] Discuss potential improvements 
- [x] Verify that the weights are still in 8bit after the loading (once there are more advances on Tim's side)
- [x] Add documentation (Younes first and then Tim)
- [x] Add a demo / few lines to explain how to use it
- [ ] Add flag that loads directly to 8bit @TimDettmers

Resources:

- WIP branch of bitsandbytes: https://github.com/TimDettmers/bitsandbytes/tree/cublaslt

Many thanks to @justheuristic and @TimDettmers !! 🎉

younesbelkada committed:
first commit

younesbelkada committed:
correct replace function

younesbelkada committed:
add final changes

- works like charm!
- cannot implement tests yet
- tested

younesbelkada committed:
clean up a bit

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_17901). All of your documentation changes will be reflected on that endpoint.

sgugger reviewed (commented):
I'm okay with this PR in principle and it looks very cool the way it would work for any model (that uses linear layers).

I have just one concern when `low_cpu_mem_usage=True` or `device_map is not None`, where the weights are re-loaded by setting attributes to the modules with a specific dtype, and I fear we lose the int8 at that step. So having a test that double checks this (and that the model can be executed normally) is key before we merge this.

Then we'll need to add some documentation to make sure people know this feature exists! And ideally, we'll need a released of `bitsandbytes` before merging, so it's easy for users to install.

younesbelkada commented:
cc-ing also @michaelbenayoun  in case you want to have a look as well ;)

LysandreJik commented:
Nice, thanks for working on it @younesbelkada! Also quite interested in the feature. I'd be particularly interested in seeing a bit of documentation so that we may understand better how it works under the hood and how to use the feature to its best.

Thanks!

younesbelkada committed:
add bitsandbytes dependencies

younesbelkada committed:
working version

- added import function
- added bitsandbytes utils file

younesbelkada commented:
Hi all!
Just to summarise a bit about what is happening and the solution we came up to implement this!
In the previous version, we found out 2 major bugs: 1- the function `set_module_tensor_to_device` seems to overwrite the `Int8Params` modules by `nn.Parameter` modules. 2-  `init_empty_weights` seems also to replace the `Int8Params` modules by `nn.Parameter` modules.

I see two solutions to this
1- Open a PR in `accelerate` to support the correct overwriting into `Int8Params` class as the following: https://github.com/huggingface/accelerate/compare/main...TimDettmers:accelerate:integration_8bit - only 2 functions are modified and should not break backward compatibility
2- Manually redefine the functions `set_module_tensor_to_device` and `init_empty_weights`as two new function `set_module_8bit_tensor_to_device` and `init_8bit_empty_weights` as proposed in this PR. 

I personally found the option 1 cleaner but the option 2 might be safer for `accelerate` - Let us know what do you think !
cc @LysandreJik @sgugger @TimDettmers

younesbelkada committed:
Merge branch 'main' of https://github.com/huggingface/transformers into integration-8bit

younesbelkada committed:
small fix

younesbelkada committed:
small fix

- fix import issue

younesbelkada committed:
fix import issues

sgugger commented on src/transformers/__init__.py:
```diff
@@ -51,6 +51,7 @@
 # Base objects, independent of any specific backend
 _import_structure = {
     "benchmark": [],
+    "bitsandbytes_utils": [],
```
Let's not make the same mistake as with lots of other utils modules and actually put it in utils/

sgugger commented on src/transformers/bitsandbytes_utils.py:
```diff
@@ -0,0 +1,149 @@
+
+def set_module_8bit_tensor_to_device(module, tensor_name, device, value=None):
+    """
+    Args:
```
This should be before the args are introduced.

younesbelkada replied:
Should be fixed in 11555491784bf4912aadbce7c9cf641f0c3bd9b9

sgugger reviewed (commented):
I'm having trouble understanding your code as I don't see any parameters inside a model that has this attribute `has_fp16_weights`. Where does this come from?

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

younesbelkada commented:
Thank you very much for your comments!
`has_fp16_weights` comes from the class `bnb.Int8Params` that is currently being developed in a WIP branch that should be merged soon on the main branch of `bitsandbytes`. Basically the logic behind it is that if the module contain this attribute then it has to be a `bnb.Int8Params` module.
I will refactor the code with your proposed changes and ask for a second batch of review 🚀

younesbelkada committed:
refactor a bit

- move bitsandbytes utils to utils
- change comments on functions

younesbelkada committed:
Merge branch 'integration-8bit' of https://github.com/younesbelkada/transformers into integration-8bit

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -554,7 +559,10 @@ def _load_state_dict_into_meta_model(
+            if not load_in_8bit:
+                set_module_tensor_to_device(model, param_name, param_device, value=param)
+            else:
+                set_module_8bit_tensor_to_device(model, param_name, param_device, value=param)
```
should be fixed in 11555491784bf4912aadbce7c9cf641f0c3bd9b9

younesbelkada commented on src/transformers/bitsandbytes_utils.py:
```diff
@@ -0,0 +1,149 @@
+        module (`torch.nn.Module`): The module in which the tensor we want to move lives. param_name (`str`): The full
+        name of the parameter/buffer. device (`int`, `str` or `torch.device`): The device on which to set the tensor.
+        value (`torch.Tensor`, *optional*): The value of the tensor (useful when going from the meta device to any
+            other device).
```
I proposed a fix in 11555491784bf4912aadbce7c9cf641f0c3bd9b9

younesbelkada replied:
Actually no, let me repush something

younesbelkada committed:
reformat docstring

- reformat docstring on init_empty_weights_8bit

sgugger commented on src/transformers/__init__.py:
```diff
@@ -453,6 +453,7 @@
         "is_vision_available",
         "logging",
     ],
+    "utils.bitsandbytes_utils": [],
```
```suggestion
    "utils.bitsandbytes": [],
```
No need to repeat `utils`.

younesbelkada replied:
fixed in 39c46a07648fa5565117c6ac33bfcbe69449f2c0

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -2063,12 +2075,18 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
 
             logger.info("Detected DeepSpeed ZeRO-3: activating zero.init() for this model")
             init_contexts = [deepspeed.zero.Init(config_dict_or_path=deepspeed_config())] + init_contexts
+        elif load_in_8bit:
+            init_contexts = [init_empty_weights_8bit()]  # Force enable init empty weights
+            logger.info("Detected 8-bit loading: activating 8-bit loading for this model")
```
I don't understand why we need a different function here, as the linear layers have not been replaced yet.

younesbelkada replied:
Actually my bad, these functions are needed. As far as I understood from my experiments a bug occurs there because `init_empty_weights` force assign the weights to be `nn.Parameter` when recursing on all modules, including the `Linear8bitLt` layers. So at the moment we put the modules on the meta device, we have to make sure their parameters are from the class `bnb.nn.Int8Params` I think

younesbelkada replied:
So we do not need it here specifically but will need it inside the `replace_8bit` function !

sgugger replied:
At which point do we have `Linear8bitLt` layers since the conversion happens after the initialization?

younesbelkada replied:
So for now it happens here: https://github.com/younesbelkada/transformers/blob/b92c25c3835f37667ee033a2b406846da06f16f4/src/transformers/utils/bitsandbytes.py#L91

younesbelkada replied:
And planning to use the `init_empty_weights_8bit` context manager there

younesbelkada replied:
@sgugger there might be a simpler option: by replacing [theses](https://github.com/huggingface/accelerate/blob/70ca65a9a1c336fe02800a0bb319ccbf4b9c6d26/src/accelerate/big_modeling.py#L68) lines by:
```
param_cls = type(module._parameters[name])
kwargs = module._parameters[name].__dict__
module._parameters[name] = param_cls(module._parameters[name].to(torch.device(“meta”)), **kwargs)
```
everything works smoothly, wdyt about opening a PR there?

sgugger replied:
Works for me as long as it works for the special subclasses used by bitsandbytes!

sgugger commented on src/transformers/utils/bitsandbytes_utils.py:
```diff
@@ -0,0 +1,158 @@
+
+
+@contextmanager
+def init_empty_weights_8bit(include_buffers=False):
```
As said above, we shouldn't need to change the decorator for `init_empty_weights` as the linear layers have not been replaced yet, so there is no parameter that will have `has_fp16_weights` attributes.

younesbelkada replied:
I think that you are right, let me double check, we might not need this function at all

younesbelkada replied:
I can confirm that you are right, proposed a modification in b92c25c3835f37667ee033a2b406846da06f16f4

younesbelkada committed:
Update src/transformers/__init__.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

younesbelkada committed:
revert bad formatting

younesbelkada commented on src/transformers/__init__.py:
```diff
@@ -118,8 +118,8 @@
     "models": [],
-    # Models
     "models.albert": ["ALBERT_PRETRAINED_CONFIG_ARCHIVE_MAP", "AlbertConfig"],
+    # Models
```
should be fixed in e116e21b4d708c8a4bfb70566e0c7ec14416dc48

younesbelkada committed:
change to bitsandbytes

younesbelkada committed:
refactor a bit

- remove init8bit since it is useless

younesbelkada committed:
more refactoring

- fixed init empty weights issue
- added threshold param

younesbelkada commented:
I think before merging we need:
- [ ] Memory footprint benchmarking
- [ ] Infrence speed benchmarking
- [ ] `lm-eval` benchmarking for large models (it has been done for small models)
- [ ] Merging the WIP branch of `bitsandbytes` into `main`

younesbelkada committed:
small hack to make it work

younesbelkada commented:
Added another PR to support `accelerate` on multi-GPU setup here: https://github.com/huggingface/accelerate/pull/539 !

younesbelkada committed:
Merge branch 'main' into integration-8bit

younesbelkada committed:
Update src/transformers/modeling_utils.py

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -1779,9 +1785,10 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
+        offload_state_dict = kwargs.pop("offload_state_dict", False)
+        load_in_8bit = kwargs.pop("load_in_8bit", False)
+        int8_threshold = kwargs.pop("int8_threshold", None)
+subfolder = kwargs.pop("subfolder", "")
```
```suggestion
        subfolder = kwargs.pop("subfolder", "")
```

younesbelkada committed:
Update src/transformers/modeling_utils.py

younesbelkada committed:
revmoe the small hack

younesbelkada committed:
Merge branch 'integration-8bit' of https://github.com/younesbelkada/transformers into integration-8bit

younesbelkada committed:
modify utils file

younesbelkada committed:
make style + refactor a bit

younesbelkada committed:
create correctly device map

younesbelkada committed:
add correct dtype for device map creation

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -1779,9 +1785,10 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
-        offload_state_dict = kwargs.pop("offload_state_dict", None)
+        offload_state_dict = kwargs.pop("offload_state_dict", False)
+        load_in_8bit = kwargs.pop("load_in_8bit", False)
+        int8_threshold = kwargs.pop("int8_threshold", None)
```
```suggestion
        int8_threshold = kwargs.pop("int8_threshold", 6.0)
```
We set 6.0 later on so let's use that here.

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -1801,6 +1808,15 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
+                    "Using `load_in_8bit=True` requires Accelerate: `pip install accelerate` and the latest version of"
+                    " bitsandbytes `pip install bitsandbytes`"
+                )
+            if not int8_threshold:
+                int8_threshold = 6.0
```
```suggestion
```
See above.

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -1779,9 +1785,10 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
         offload_folder = kwargs.pop("offload_folder", None)
-        offload_state_dict = kwargs.pop("offload_state_dict", None)
+        offload_state_dict = kwargs.pop("offload_state_dict", False)
+        load_in_8bit = kwargs.pop("load_in_8bit", False)
+        int8_threshold = kwargs.pop("int8_threshold", None)
```
Those two new kwargs should be documented.

younesbelkada replied:
Thanks ! Should be fixed in a5cd157b03311d99661c27ea72e6348406ffdccd

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -2159,7 +2185,7 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
 
         # Dispatch model with hooks on all devices if necessary
-        if device_map is not None:
+        if device_map:
```
Please revert, we always explicitly test instead of relying on Python bool magic.

younesbelkada replied:
Great !  Should be fixed in a4c19c193b598efae80f2489298c3aeeacb50804

sgugger commented on src/transformers/utils/bitsandbytes.py:
```diff
@@ -0,0 +1,98 @@
+    Args:
+        module (`torch.nn.Module`):
+            The module in which the tensor we want to move lives.
+        tensor_name (`str`): The full
+            name of the parameter/buffer.
```
Weird formatting.

younesbelkada replied:
Thanks for spotting this ;) ! Fixed in 9bd326b7eaf758e19f12166ec0bbba1906d53062

sgugger commented on src/transformers/utils/bitsandbytes.py:
```diff
@@ -0,0 +1,98 @@
+    if has_fp16_weights is not None:
+        param = module._parameters[tensor_name]
+        if param.device.type != "cuda":
+            with torch.no_grad():
```
Why is this necessary, none of the operations below imply some gradient computation.

younesbelkada replied:
Yes can confirm you are right, should be fixed in a4c19c193b598efae80f2489298c3aeeacb50804

sgugger commented on src/transformers/utils/bitsandbytes.py:
```diff
@@ -0,0 +1,98 @@
+                )
+                module._parameters[tensor_name] = new_value
+    else:
+        with torch.no_grad():
```
Same here.

younesbelkada replied:
Fixed in a4c19c193b598efae80f2489298c3aeeacb50804 !

sgugger commented on src/transformers/utils/bitsandbytes.py:
```diff
@@ -0,0 +1,98 @@
+
+
+def replace_8bit_linear(model, threshold=6.0):
+
```
This would benefit from a docstring so that future code readers know exactly what this function does :-)

younesbelkada replied:
Great! Fixed in 9bd326b7eaf758e19f12166ec0bbba1906d53062

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

younesbelkada committed:
apply suggestions

- remove with torch.grad
- do not rely on Python bool magic!

younesbelkada committed:
add docstring

 - add docstring for new kwargs

younesbelkada committed:
add docstring

- comment `replace_8bit_linear` function
- fix weird formatting

younesbelkada commented:
Thanks @sgugger for your review ! Fixed the suggestions ;) 
I think that we are good to go to merge https://github.com/huggingface/accelerate/pull/539 if you don't mind 🙏 
I just need to wait the release of `bitsandbytes` to be more stable (facing some issues when installing the library but should be fixed very soon). Once this is fixed I think that we should be good to go for merging 🚀

sgugger commented:
Merged the PR in Accelerate! Don't forget to add some documentation and also setup some tests for this so it doesn't great broken by future PRs :-)

younesbelkada commented:
- [ ] Have a working colab demo for inference
- [ ] Add more documentation
- [ ] Implement tests

younesbelkada committed:
- added more documentation
- added new utility function for memory footprint tracking
- colab demo to add

younesbelkada committed:
few modifs

- typo doc
- force cast into float16 when load_in_8bit is enabled

younesbelkada committed:
added colab link

younesbelkada committed:
add test architecture + docstring a bit

younesbelkada committed:
refactor a bit testing class

younesbelkada committed:
make style + refactor a bit

younesbelkada committed:
enhance checks

- add more checks
- start writing saving test

younesbelkada committed:
Merge branch 'integration-8bit' of https://github.com/younesbelkada/transformers into integration-8bit

younesbelkada committed:
clean up a bit

younesbelkada committed:
male style

younesbelkada committed:
add more details on doc

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -1845,10 +1845,18 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
+                    "A device map needs to be passed to run convert models into mixed-int8 format. Please run"
+                    "`.from_pretrained` with `device_map='auto'`"
+                )
+            if (from_tf | from_flax):
```
```suggestion
            if (from_tf or from_flax):
```
Let's use the English version which is more readable ;-)

younesbelkada replied:
Yeah ahah thanks for the suggestion!

younesbelkada replied:
I could not commit the suggestion from the GUI, therefore pushed a commit here: f6eb9456a8459cec5774bfa22b377c7e0ffc859f

younesbelkada committed:
add more tests

- still needs to fix 2 tests

younesbelkada committed:
replace by "or"

- could not fix it from GitHub GUI

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

younesbelkada committed:
refactor a bit testing code + add readme

younesbelkada committed:
make style

younesbelkada committed:
fix import issue

younesbelkada commented:
Before moving forward, I would like to have a comment from @michaelbenayoun @mfuntowicz and @echarlaix 

## About this PR

We replace all the `nn.Linear` modules by the `bnb.Linear8bitLt` modules from the recent release of `bitsandbytes` that proposes a new post-training quantization technique for 0 performance degradation on large-scale models (>1b parameters). With that we have managed to fit BLOOM-176B on 3xA100 80GB instead of 6xA100 GB with no performance degradation. 

## About the mixed quantization method in few words

In this technique the operations on the outliers are done in `fp16` and the rest of the operations are done in `int8` to achieve 0 performance degradation on large-scale models. 

## Usage

This does not run on CPU, you will need a GPU that supports 8-bit core tensors operations (T4 and A100) to make it run. Here is a tutorial on Google Colab on how to run the mixed-int8 model: https://colab.research.google.com/drive/1qOjXfQIAULfKvZqwCen8-MoWKGdSatZ4#scrollTo=YJlldexxwnhM

michaelbenayoun commented on src/transformers/modeling_utils.py:
```diff
@@ -1691,6 +1715,16 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
+                https://test.pypi.org/simple/ bitsandbytes-cudaXXX` where XXX is your CUDA version (e.g. 11.6 = 116).
+                Make also sure that you have enough GPU RAM to store half of the model size since the 8bit modules are
+                not compiled and adapted for CPUs.
+            int8_threshold (`int`, *optional*):
```
Maybe give more details here on what it actually means so that the users do not have to read the paper to be able to use it.

younesbelkada replied:
Thank you for pointing out this! 
I added more description on 7abb914e726e4f4d3845d4faed2c0085b5e73e13 and 961e57ea85d943b0316f9f69e43dff6e7f0915a4 that should let user understand the usage of this parameter.

TimDettmers replied:
If you want to add more information, you could write:
>Values are usually normally distributed, that is, most values are in the range [-3.5, 3.5], but there are some exceptional systematic outliers that are very differently distributed for large models. These outliers are often in the interval [-60, -6] or [6, 60]. Int8 quantization works well for values of magnitude ~5, but beyond that, there is a significant performance penalty. A good default threshold is 6, but a lower threshold might be needed for more unstable models (small models, fine-tuning).

younesbelkada replied:
Nice thanks! Added some suggested comments here: 0324f4b1885e51cc93f3f25754b1469f7538bbc7

michaelbenayoun commented on src/transformers/modeling_utils.py:
```diff
@@ -1801,6 +1836,30 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
+            if from_tf or from_flax:
+                raise ValueError(
+                    "Converting into mixed 8-bit weights from tf/flax weights is currently not supported, please make"
+                    " surethe weights are in PyTorch format."
```
```suggestion
                    " sure the weights are in PyTorch format."
```

michaelbenayoun commented on src/transformers/utils/bitsandbytes.py:
```diff
@@ -0,0 +1,112 @@
+        value (`torch.Tensor`, *optional*):
+            The value of the tensor (useful when going from the meta device to any other device).
+    """
+    # Recurse if needed
+    if "." in tensor_name:
+        splits = tensor_name.split(".")
+        for split in splits[:-1]:
+            new_module = getattr(module, split)
+            if new_module is None:
+                raise ValueError(f"{module} has no attribute {split}.")
+            module = new_module
+        tensor_name = splits[-1]
+
+    if tensor_name not in module._parameters and tensor_name not in module._buffers:
+        raise ValueError(f"{module} does not have a parameter or a buffer named {tensor_name}.")
+    is_buffer = tensor_name in module._buffers
+    old_value = getattr(module, tensor_name)
```
Maybe I am wrong, but I think you can make it simpler:
```
if tensor_name in [t[0] for t in module.named_parameters()]:
    old_value = module.get_parameter(tensor_name)
elif tensor_name in [t[0] for t in module.named_buffers()]:
    old_value = module.get_buffer(tensor_name)
else:
    raise ValueError ...
```

michaelbenayoun commented on src/transformers/utils/bitsandbytes.py:
```diff
@@ -0,0 +1,112 @@
+            `int8_threshold` for outlier detection as described in the formentioned paper. This parameters is set to
+            `6.0` as described by the paper.
+    """
+    for n, module in model.named_children():
```
I think this could be better written as a `torch.fx` transformation, but this way works for more models since not all models are currently traceable for torch.fx.

michaelbenayoun reviewed (commented):
Vert nice!
I think we should ask ourselves if this should be merged here or in `optimum`.

younesbelkada committed:
Update src/transformers/modeling_utils.py

Co-authored-by: Michael Benayoun <mickbenayoun@gmail.com>

younesbelkada committed:
add few comments

younesbelkada committed:
add more doctring + make style

younesbelkada committed:
more docstring

younesbelkada committed:
raise error when loaded in 8bit

younesbelkada committed:
make style

younesbelkada commented:
Can confirm the slow tests that I have designed are passing on my testing machine (Tesla T4 15GB). But for now it is not possible to load saved int8 checkpoints because you need to load the quantization statistics that are not saved when doing `model.state_dict()` in `bitsandbytes`. For now I propose to just raise an error message if int8 weights are loaded and tell users that the feature is not supported (as proposed in 1326a42795033410dae6c5a8a07b81f12ee7a41c). 
No strong opinions but I personally advocate to keep this feature inside `transformers` since the method relies also on `accelerate` + an additional lib (`bitsandbytes`), but I am not the best knowledgable person regarding `optimum` integration that might be a bit different than the `transformers` one. 
cc @sgugger @mfuntowicz @TimDettmers🙏

TimDettmers commented:
Thank you for all the work on this PR @younesbelkada, @sgugger, @michaelbenayoun! 

Regarding the `transformers` vs `optimum` question: From my understanding of the libraries, I think if people want to deploy models or run them with high efficiency `optimum` seems to be the right tool, whereas general purpose "inefficient" access of models is more suitable for `transformers`. 

As such, I think it's best to keep this feature in `transformers`. I think it fits better into there since it is not meant for fast inference but memory-efficient inference for as many use-cases as possible.

TimDettmers commented on tests/mixed_int8/README.md:
````diff
@@ -0,0 +1,13 @@
+```conda create --name int8-testing python==3.8```
+```git clone https://github.com/younesbelkada/transformers.git && git checkout integration-8bit```
+```pip install -e ".[dev]"```
+```pip install -i https://test.pypi.org/simple/ bitsandbytes-cuda114```
````
I just changed the way bitsandbytes is installed. This line no longer uses the cuda suffix: 
```pip install -i https://test.pypi.org/simple/ bitsandbytes```

TimDettmers commented on src/transformers/utils/bitsandbytes.py:
```diff
@@ -0,0 +1,117 @@
+
+def replace_8bit_linear(model, threshold=6.0):
+    """
+    A helper function to replace all `torch.nn.Linear` modules by `bnb.nn.Linear8bit` modules from the `bitsandbytes`
+    library. This will enable running your models using mixed int8 precision as described by the paper `GPT3.int8():
+    8-bit Matrix Multiplication for Transformers at Scale`. Make sure `bitsandbytes` compiled with the correct CUDA
+    version of your hardware is installed before running this function. `pip install -i https://test.pypi.org/simple/
+    bitsandbytes-cudaXXX` with `XXX` is your CUDA version (e.g., 11.6 = 116)
+
+    The function will be run recursively and replace all `torch.nn.Linear` modules except for the `lm_head` that should
+    be kept as a `torch.nn.Linear` module. The replacement is done under `init_empty_weights` context manager so no
+    CPU/GPU memory is required to run this function.
```
The new release now works with just a pip install. And it is no longer needed to compile bitsandbytes. For the full release (done in a couple of days) the library would also reside directly on pip. So best to change this already to: `pip install bitsandbytes`

It might be useful to add one line about how it works before linking to the GPT3.int8() paper:
>Int8 mixed-precision matrix decomposition works by separating a matrix multiplication into two streams: (1) and systematic feature outlier stream matrix multiplied in fp16 (0.01%), (2) a regular stream of int8 matrix multiplication (99.9%). With this method, int8 inference with no predictive degradation is possible for very large models (>=176B parameters).

TimDettmers commented on docs/source/en/main_classes/model.mdx:
````diff
@@ -133,6 +133,31 @@ model = AutoModel.from_config(config)
+- Install `accelerate`:
+`pip install accelerate`
+
+#### Running mixed-int8 models
+
+After carefully installing the required libraries, the way to load your mixed 8-bit model is as follows:
+```
+model_name = "bigscience/bloom-2b5"
+model_8bit = AutoModelForCausalLM.from_pretrained(model_name, device_map="auto", load_in_8bit=True)
+```
+The implementation supports multi-GPU setup thanks to `accelerate` as backend. 
````
I think it is important to add the `max_memory` flag with an example (for example 2 GPUs). A friend had problems running models because he did not know how to avoid the out of memory errors he was getting.

younesbelkada replied:
Great agrred! Should added in 0324f4b1885e51cc93f3f25754b1469f7538bbc7 ;)

TimDettmers commented on docs/source/en/main_classes/model.mdx:
```diff
@@ -133,6 +133,31 @@ model = AutoModel.from_config(config)
+
+- Make sure you run that on a NVIDIA T4 or A100 GPU that supports 8-bit tensor cores. Note that previous generations of NVIDIA GPUs do not support 8-bit tensor cores.
+- Install the correct version of `bitsandbytes` by running:
+`pip install -i https://test.pypi.org/simple/ bitsandbytes-cudaXXX` with `XXX` corresponding to your `CUDA` version. E.g, if you have `CUDA` 11.3 installed on your machine, run `pip install -i https://test.pypi.org/simple/ bitsandbytes-cuda113`
```
This also needs to be updated to just `pip install bitsandbytes`

TimDettmers commented on docs/source/en/main_classes/model.mdx:
```diff
@@ -133,6 +133,31 @@ model = AutoModel.from_config(config)
 
+### Mixed int8 quantization 🪶
+
+From the paper `GPT3.int8() : 8-bit Matrix Multiplication for Transformers at Scale`, we suport HuggingFace 🤗  integration for all models in the Hub with few lines of code. 
+For models trained in  half-precision (aka, either `float16` or `bfloat16`). This method aims to reduce `nn.Linear` size by 2, without affecting too much quality by operating on the outliers in half-precision. Therefore the method is termed as mixed 8bit quantization.
+This technique is useful and works well for billion scale models (>1B parameters) therefore we advice you to use it only for models of that scale. This method has been tested for 2-billion to 176-billion scale models and supports only PyTorch models. 
+Note also that you would require a GPU to run mixed-8bit models as the kernels has been compiled for GPUs only. Make sure that you have enough GPU RAM to store the quarter (or half if your model is natively in half precision) of the model before using this feature. 
+
+Below are some notes to help you use this module, or follow this demo on Google colab: [![Open In Google Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1qOjXfQIAULfKvZqwCen8-MoWKGdSatZ4?usp=sharing)
```
Maybe add some information from the other comments that I left

younesbelkada replied:
Yes agreed here too! Added some docs here: 0324f4b1885e51cc93f3f25754b1469f7538bbc7

TimDettmers reviewed (commented):
Great work, @younesbelkada! I think the most important part is to add `max_memory` flag to the default example (best would be to use an example with 2 GPUs, max memory 20GB when the GPUs have 24GB). 

After that just updating and extending some docs. Otherwise, everything looks good to me!

younesbelkada committed:
add warning if loaded on CPU

younesbelkada committed:
add small sanity check

ydshieh commented on utils/tests_fetcher.py:
```diff
@@ -464,6 +464,7 @@ def module_to_test_file(module_fname):
     "tests/sagemaker/test_single_node_gpu.py",  # SageMaker test
     "tests/sagemaker/test_multi_node_model_parallel.py",  # SageMaker test
     "tests/sagemaker/test_multi_node_data_parallel.py",  # SageMaker test
+    "tests/mixed_int8/test_mixed_int8.py",  # SageMaker test
```
Should not be `# SageMaker test` ..?

younesbelkada replied:
Yes thank you for spotting this! Should be fixed in: 6d93424992ffdbfc57a8e944344a4d80633a97f1

younesbelkada committed:
fix small comment

younesbelkada committed:
add bitsandbytes on dockerfile

younesbelkada committed:
Improve documentation

- improve documentation from comments

younesbelkada committed:
Merge branch 'main' into integration-8bit

younesbelkada committed:
add few comments

younesbelkada committed:
slow tests pass on the VM but not on the CI VM

younesbelkada committed:
Merge branch 'integration-8bit' of https://github.com/younesbelkada/transformers into integration-8bit

younesbelkada committed:
Fix merge conflict

younesbelkada committed:
make style

cnbeining commented:
Forgive me for jumping the gun - 

On Colab(T4, 12G RAM) I tried:

```
!nvidia-smi

| NVIDIA-SMI 460.32.03    Driver Version: 460.32.03    CUDA Version: 11.2     |
|   0  Tesla T4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   58C    P8    10W /  70W |      0MiB / 15109MiB |      0%      Default |
|  No running processes found                                                 |
```

Then

```
!pip install https://github.com/younesbelkada/transformers/archive/refs/heads/integration-8bit.zip accelerate 
!pip install -i https://test.pypi.org/simple/ bitsandbytes-cuda112
```

Loading model with

```
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM

tokenizer = AutoTokenizer.from_pretrained("Salesforce/codegen-2B-mono")
model = AutoModelForCausalLM.from_pretrained("Salesforce/codegen-2B-mono", load_in_8bit=True, device_map="auto")
```

And got this error:

```
RuntimeError                              Traceback (most recent call last)
[<ipython-input-8-40073518cc86>](https://localhost:8080/#) in <module>()
----> 1 model = AutoModelForCausalLM.from_pretrained("Salesforce/codegen-2B-mono", load_in_8bit=True, device_map="auto")

7 frames
[/usr/local/lib/python3.7/dist-packages/transformers/models/auto/auto_factory.py](https://localhost:8080/#) in from_pretrained(cls, pretrained_model_name_or_path, *model_args, **kwargs)
    444         elif type(config) in cls._model_mapping.keys():
    445             model_class = _get_model_class(config, cls._model_mapping)
--> 446             return model_class.from_pretrained(pretrained_model_name_or_path, *model_args, config=config, **kwargs)
    447         raise ValueError(
    448             f"Unrecognized configuration class {config.__class__} for this kind of AutoModel: {cls.__name__}.\n"

[/usr/local/lib/python3.7/dist-packages/transformers/modeling_utils.py](https://localhost:8080/#) in from_pretrained(cls, pretrained_model_name_or_path, *model_args, **kwargs)
   2282         # Dispatch model with hooks on all devices if necessary
   2283         if device_map is not None:
-> 2284             dispatch_model(model, device_map=device_map, offload_dir=offload_folder)
   2285 
   2286         if output_loading_info:

[/usr/local/lib/python3.7/dist-packages/accelerate/big_modeling.py](https://localhost:8080/#) in dispatch_model(model, device_map, main_device, state_dict, offload_dir, offload_buffers, preload_module_classes)
    246         offload_buffers=offload_buffers,
    247         weights_map=weights_map,
--> 248         preload_module_classes=preload_module_classes,
    249     )
    250     model.hf_device_map = device_map

[/usr/local/lib/python3.7/dist-packages/accelerate/hooks.py](https://localhost:8080/#) in attach_align_device_hook_on_blocks(module, execution_device, offload, weights_map, offload_buffers, module_name, preload_module_classes)
    446             place_submodules=True,
    447         )
--> 448         add_hook_to_module(module, hook)
    449         attach_execution_device_hook(module, execution_device[module_name])
    450     elif module_name in execution_device:

[/usr/local/lib/python3.7/dist-packages/accelerate/hooks.py](https://localhost:8080/#) in add_hook_to_module(module, hook)
    136         module._old_forward = old_forward
    137 
--> 138     module = hook.init_hook(module)
    139     module._hf_hook = hook
    140 

[/usr/local/lib/python3.7/dist-packages/accelerate/hooks.py](https://localhost:8080/#) in init_hook(self, module)
    219         if not self.offload and self.execution_device is not None:
    220             for name, _ in named_module_tensors(module, recurse=self.place_submodules):
--> 221                 set_module_tensor_to_device(module, name, self.execution_device)
    222         elif self.offload:
    223             self.original_devices = {

[/usr/local/lib/python3.7/dist-packages/accelerate/utils/modeling.py](https://localhost:8080/#) in set_module_tensor_to_device(module, tensor_name, device, value)
    128         module._buffers[tensor_name] = new_value
    129     else:
--> 130         new_value = nn.Parameter(new_value, requires_grad=old_value.requires_grad)
    131         module._parameters[tensor_name] = new_value
    132 

[/usr/local/lib/python3.7/dist-packages/torch/nn/parameter.py](https://localhost:8080/#) in __new__(cls, data, requires_grad)
     40         t = data.detach().requires_grad_(requires_grad)
     41         if type(t) is not type(data):
---> 42             raise RuntimeError(f"Creating a Parameter from an instance of type {type(data).__name__} "
     43                                "requires that detach() returns an instance of the same type, but return "
     44                                f"type {type(t).__name__} was found instead. To use the type as a "

RuntimeError: Creating a Parameter from an instance of type Int8Params requires that detach() returns an instance of the same type, but return type Tensor was found instead. To use the type as a Parameter, please correct the detach() semantics defined by its __torch_dispatch__() implementation.
```

Interestingly on AWS Sagemaker(T4, 16G RAM) -

```
!pip install -i https://test.pypi.org/simple/ bitsandbytes-cuda114

import torch
from transformers import AutoTokenizer, AutoModelForCausalLM
tokenizer = AutoTokenizer.from_pretrained("Salesforce/codegen-6B-mono")
model = AutoModelForCausalLM.from_pretrained("Salesforce/codegen-6B-mono", load_in_8bit=True, device_map="auto")
```

got me

```
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
/tmp/ipykernel_141/3855166932.py in <cell line: 1>()
----> 1 model = AutoModelForCausalLM.from_pretrained("Salesforce/codegen-6B-mono", load_in_8bit=True, device_map="auto")

~/.conda/envs/default/lib/python3.9/site-packages/transformers/models/auto/auto_factory.py in from_pretrained(cls, pretrained_model_name_or_path, *model_args, **kwargs)
    444         elif type(config) in cls._model_mapping.keys():
    445             model_class = _get_model_class(config, cls._model_mapping)
--> 446             return model_class.from_pretrained(pretrained_model_name_or_path, *model_args, config=config, **kwargs)
    447         raise ValueError(
    448             f"Unrecognized configuration class {config.__class__} for this kind of AutoModel: {cls.__name__}.\n"

~/.conda/envs/default/lib/python3.9/site-packages/transformers/modeling_utils.py in from_pretrained(cls, pretrained_model_name_or_path, *model_args, **kwargs)
   2177             init_contexts = [deepspeed.zero.Init(config_dict_or_path=deepspeed_config())] + init_contexts
   2178         elif load_in_8bit:
-> 2179             init_contexts = [init_empty_weights()]  # Force enable init empty weights
   2180             logger.info("Detected 8-bit loading: activating 8-bit loading for this model")
   2181         elif low_cpu_mem_usage:

NameError: name 'init_empty_weights' is not defined
```

I suppose the 2nd case may have something to do with environment setup - but what would trigger the first issue?

Thanks,

younesbelkada commented:
Hi @cnbeining !
Thanks for your interest in this feature and happy to see that you are already excited to run it on Codegen! 🚀 
Initially your problem is related to `accelerate` that you are installing. Make sure you install the latest version from source using a command like:
```
pip install git+https://github.com/huggingface/accelerate.git@24c28a1adc284db0126b7c17ebef275597ddc6b7
```
With 24c28a1adc284db0126b7c17ebef275597ddc6b7 being the latest commit hash from accelerate. The most recent release (aka `accelerate` library that you will get from `pip install accelerate`) is not compatible with this PR. Therefore you will need the latest version of it.

However, when using `load_in_8bit`, `torch_dtype=torch.float16` is internally called. It happens that there might be a small bug in Codegen when using `torch_dtype=torch.float16` that we propose to fix in https://github.com/huggingface/transformers/pull/18467 . if you are interested to reproduce the issue you can run this small snippet:

```
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM

tokenizer = AutoTokenizer.from_pretrained("Salesforce/codegen-2B-mono")
model = AutoModelForCausalLM.from_pretrained("Salesforce/codegen-2B-mono", device_map="auto", torch_dtype=torch.float16)

text = "def quicksort(l):"

encoded_input = tokenizer(text, return_tensors='pt')
output_sequences = model.generate(input_ids=encoded_input['input_ids'], attention_mask=encoded_input['attention_mask'])
print(tokenizer.decode(output_sequences[0], skip_special_tokens=True))
```

Since this might take time to be merged and I saw that you wanted to run on Google Colab I made a special branch that you can build from Colab and should work (tested it) here: [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1IUAn97Zsfiiz7B1-vAAOSGsWE96yAiNM#scrollTo=1mNklAh5trGY). Just run those cells and everything should work.

If you follow the same installation instructions as this Colab I think that everything should work smoothly in SageMaker as well but we never know! 
Let us know if this helps, and happy to help you again if necessary 💪 
Also if you face any other issues, I think that it would be better to move discussion into an issue! 🐛 

Thanks
Younes

younesbelkada committed:
another test should pass on a multi gpu setup

younesbelkada committed:
fix bad import in testing file

younesbelkada committed:
Fix slow tests

- remove dummy batches
- no more CUDA illegal memory errors

younesbelkada committed:
odify dockerfile

younesbelkada commented:
@LysandreJik I have a question regarding slow tests for this feature! 
I prefer to build another Docker image for these tests and them separately because it happens sometimes that the import of `bitsandbytes` fails on some specific configurations. We found an issue that will be fixed on `bitsandbytes` asap but I think that having a separate image and running the tests independently is safer to not affect other tests. Since `bitsandbytes` is [always being imported](https://github.com/younesbelkada/transformers/blob/31fce94e8a3983dfa65222311b340460ccff05f7/src/transformers/modeling_utils.py#L95) if it is available if the docker image installs it all tests will fail at import time. I can also try to come up with a solution where we import this library only if `load_in_8bit` is triggered. What do you think is the best in this case?

younesbelkada commented on docs/source/en/main_classes/model.mdx:
````diff
@@ -133,6 +133,41 @@ model = AutoModel.from_config(config)
+The implementation supports multi-GPU setup thanks to `accelerate` as backend. If you want to control the GPU memory you want to allocate for each GPU, you can use the `max_memory` argument as follows:
+(If allocating `1GB` into GPU-0 and `2GB` into GPU-1, you can use `max_memory={0:"1GB", 1:"2GB"}`)
+```
+max_memory_mapping={0:"2GB", 1:"3GB"}
````
```suggestion
max_memory_mapping={0:"1GB", 1:"2GB"}
```

younesbelkada committed:
Update docs/source/en/main_classes/model.mdx

younesbelkada committed:
Update Dockerfile

younesbelkada committed:
Update model.mdx

younesbelkada committed:
Update Dockerfile

younesbelkada commented:
Tests are now [all passing](https://github.com/huggingface/transformers/actions/runs/2816407110) with the latest fix of `bitsandbytes`
I would love to have a potential final round of review! cc @sgugger @LysandreJik

younesbelkada commented on tests/mixed_int8/README.md:
````diff
@@ -0,0 +1,37 @@
+```conda create --name int8-testing python==3.8```
+```git clone https://github.com/younesbelkada/transformers.git && git checkout integration-8bit```
+```pip install -e ".[dev]"```
+```pip install -i https://test.pypi.org/simple/ bitsandbytes-cuda114```
````
```suggestion
```pip install -i https://test.pypi.org/simple/ bitsandbytes```
```

younesbelkada commented on docs/source/en/main_classes/model.mdx:
```diff
@@ -133,6 +133,43 @@ model = AutoModel.from_config(config)
+
+- Make sure you run that on a NVIDIA T4 or A100 GPU that supports 8-bit tensor cores. Note that previous generations of NVIDIA GPUs do not support 8-bit tensor cores.
+- Install the correct version of `bitsandbytes` by running:
+`pip install -i https://test.pypi.org/simple/ bitsandbytes-cudaXXX` with `XXX` corresponding to your `CUDA` version. E.g, if you have `CUDA` 11.3 installed on your machine, run `pip install -i https://test.pypi.org/simple/ bitsandbytes-cuda113`
```
```suggestion
`pip install -i https://test.pypi.org/simple/ bitsandbytes-cuda`
```

younesbelkada committed:
Apply suggestions from code review

younesbelkada committed:
few modifications

- lm head can stay on disk/cpu
- change model name so that test pass

younesbelkada committed:
Merge branch 'main' into integration-8bit

younesbelkada committed:
change test value

- change test value to the correct output
- torch bmm changed to baddmm in bloom modeling when merging

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into integration-8bit

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -2062,12 +2127,18 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
             model = cls(config, *model_args, **model_kwargs)
 
+        if load_in_8bit:
+            model = replace_8bit_linear(model, threshold=int8_threshold)
```
Or the logger.info could go here actually!

younesbelkada replied:
Yes! Thanks for the proposal!
Resolving this because of: a92564137ec21e665c8e241251d5a2c565113b20

younesbelkada committed:
modify installation guidelines

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -2090,9 +2161,19 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
 
+            if load_in_8bit:
+                # The LM head can stay on disk / CPU
+                device_map_without_lm_head = {key: device_map[key] for key in device_map.keys() if key != "lm_head"}
```
This is a bit too adhoc here (though I understand this is a bit anecdotic). Some models may call it by a different name. You should rely on the methods of `PreTrainedModel` that give you the output embeddings here, probably.

younesbelkada replied:
I have the same comment as below, I am proposing a solution in 279b8c45bcf3893e1b4695d32ff2e1083c6bda9d which should be clear for that

sgugger commented on src/transformers/utils/bitsandbytes.py:
```diff
@@ -0,0 +1,120 @@
+    Parameters:
+        model (`torch.nn.Module`):
+            Input model or `torch.nn.Module` as the function is run recursively.
+        threshold (`float`, *optional*):
```
```suggestion
        threshold (`float`, *optional*, defaults to 6.0):
```

sgugger commented on src/transformers/utils/bitsandbytes.py:
```diff
@@ -0,0 +1,120 @@
+            `int8_threshold` for outlier detection as described in the formentioned paper. This parameters is set to
+            `6.0` as described by the paper.
+    """
+    for n, module in model.named_children():
```
```suggestion
    for name, module in model.named_children():
```
Better names please :-)

younesbelkada replied:
Sure ! 
I had to modify the other variables below so I propose a temporary fix here: 630b4f757557f5b2c9b0524071003f1fadb826d3

sgugger commented on src/transformers/utils/bitsandbytes.py:
```diff
@@ -0,0 +1,120 @@
+        if len(list(module.children())) > 0:
+            replace_8bit_linear(module, threshold)
+
+        if isinstance(module, nn.Linear) and n != "lm_head":
```
Here too, using `lm_head` is too adhoc IMO

younesbelkada replied:
I proposed two solutions for that: 279b8c45bcf3893e1b4695d32ff2e1083c6bda9d and e49a2ea29262a57f90c2db8c26f5655af4ad3aea ! I think that the last one is better as you suggested

younesbelkada replied:
Added more tests here: 5718d78ae2f7bb32c5afab6829e2545c9f4b06c1

sgugger reviewed (approved):
Looking great, thanks a lot! I've left a couple of comments, the most important one being to try to avoid relying on the fact the LM head is always called `lm_head`. There are utilities to:
- get the output embedding if that's what you need
- get the modules that are tied together if it's because they are tied with the embeddings.

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -1706,6 +1730,22 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
+                https://test.pypi.org/simple/ bitsandbytes-cudaXXX` where XXX is your CUDA version (e.g. 11.6 = 116).
+                Make also sure that you have enough GPU RAM to store half of the model size since the 8bit modules are
+                not compiled and adapted for CPUs.
+            int8_threshold (`int`, *optional*):
```
Oh yes you are right here, it is a float

younesbelkada replied:
```suggestion
            int8_threshold (`float`, *optional*, defaults to 6):
```

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -1823,6 +1864,30 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
+                    " bitsandbytes-cuda113` for CUDA 11.3)"
+                )
+            if torch_dtype == "auto" or torch_dtype != torch.float16:
+                torch_dtype = (
+                    torch.float16
+                )  # We force the `dtype` to be float16, this is a requirement from `bitsandbytes`
```
Great ! I was always wondering how to go around this formatting issue ahah

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

younesbelkada committed:
replace `n`by `name`

younesbelkada committed:
merge `load_in_8bit` and `low_cpu_mem_usage`

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -2062,12 +2127,18 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
 
             logger.info("Detected DeepSpeed ZeRO-3: activating zero.init() for this model")
             init_contexts = [deepspeed.zero.Init(config_dict_or_path=deepspeed_config())] + init_contexts
+        elif load_in_8bit:
```
Totally makes sense ! 
Proposed a fix in: a92564137ec21e665c8e241251d5a2c565113b20 I think that we can close this comment as well as the one below

younesbelkada committed:
first try - keep the lm head in full precision

younesbelkada committed:
better check

- check the attribute `base_model_prefix` instead of computing the number of parameters

younesbelkada committed:
added more tests

younesbelkada committed:
Update src/transformers/utils/bitsandbytes.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

younesbelkada committed:
Merge branch 'integration-8bit' of https://github.com/younesbelkada/transformers into integration-8bit

younesbelkada commented:
Going to do a last sanity check - testing with Docker and see if the slow tests passes on our Docker and merge once it's green! 🟢

cnbeining commented:
GJ!

Non blocking comment: How about incorporating (optional) `bnb.nn.StableEmbedding ` as [recommended by authors](https://github.com/facebookresearch/bitsandbytes#using-the-8-bit-optimizers) or added benefit is limited?

younesbelkada committed:
improve documentation

- fix typos for installation
- change title in the documentation

younesbelkada commented:
Thanks @cnbeining !
I think that this can be done in a separate PR since we need to release the beta version of this feature probably ASAP! Also I am not sure how the current implementation will handle tied weights if we replace Embedding layers with StableEmbedding. So this needs further tests/investigations

cnbeining commented:
Yeah let's get this rolled out to unleash GPT-J-6B and CodeGen to ordinary folks :-) I will continue with my testing with `StableEmbedding` and will report results as they come by.

Again thanks so much for all the effort!

younesbelkada commented:
Great that would be awesome! I would be definitely interested in seeing the results and comparison against the current approach (aka without StableEmbedding)
Let's maybe keep the results in this thread even after merging the PR

younesbelkada commented:
Ultimate checks are passing: https://github.com/huggingface/transformers/actions/runs/2830688091 Merging!

fxmarty commented:
Looks nice, will try it out :)

younesbelkada commented:
You can check the Google Colab: https://colab.research.google.com/drive/1qOjXfQIAULfKvZqwCen8-MoWKGdSatZ4#scrollTo=W8tQtyjp75O_ to see how to run it! We will publish that today with the beta release on Twitter

mrm8488 commented:
Great work! Can big models such us models used in the example colab be fine-tuned just loading it as `int8`?
Are you thinking about release a colab for fine-tuning a model not just for inference?
Thanks in adavace

younesbelkada commented:
Thanks for the remark @mrm8488 ! Indeed it would be very nice to have a fine-tuning demo on colab
After discussing with @TimDettmers it appears that the current implementation would support classic `torch` optimizers.
Also I think that @justheuristic has some experience with finetuning int8 models using `Linear8bitLt` modules for prompt tuning ;) so I will let him answer on the feasibility of that! 🚀

justheuristic commented:
tl;dr soon :)

@timdettmers and @dbaranchuk are working on integrating fine-tuning on the bitsandbytes.
After they're done, you'll be able to write trainable adapters, soft-prompts and other parameter-efficient methods.

There is also a group in BigScience that works on 8-bit finetuning for very large models (think bloom-176B) in colab, but we are still polishing the code. I'll tag you once it becomes public.

stas00 commented:
@younesbelkada, thank you for integrating this awesome feature - may I suggest that all these cool features will remain hidden unless we expose them in the docs where users are likely to search for those and not in the API docs.

I propose to add a new section at https://huggingface.co/docs/transformers/main/perf_train_gpu_one so that those searching for performance improvement will find it. Thank you!

younesbelkada commented:
Thanks for the comment ! Sounds really good for me 💪 
I was planning to open a PR by the beginning of next week to add the link to blogpost + paper, I will use this PR to propose your refactoring as well

pai4451 commented:
> Thanks for the comment ! Sounds really good for me 💪
> I was planning to open a PR by the beginning of next week to add the link to blogpost + paper, I will most likely use this PR to propose your refactoring as well

Hi @ younesbelkada, I want to try bitsandbytes on an 8x A6000 server (CUDA version 11.3) with BLOOM. Unfortunately, the following error throws out.

```
RuntimeError: Creating a Parameter from an instance of type Int8Params requires that detach() returns an instance of the same type, but return type Tensor. was found instead. To use the type as a Parameter, please correct the detach() semantics defined by __torch_dispatch__() implementation.
```

I use the code downloaded from Colab using the `model.generate()` way for inference instead of the pipeline from HuggingFace. Do you know how to solve the issue? I installed `bitsandbytes==0.31.8` from https://pypi.org/project/bitsandbytes/; the latest `transformers` package from the master branch also installed the latest `Accelerate` from pip.

cnbeining commented:
@pai4451 the code has not been released to PyPI [yet](https://pypi.org/project/transformers/#history) - you probably want to use `pip install git+https://github.com/huggingface/transformers.git` to get the `HEAD` that includes this PR.

pai4451 commented:
> @pai4451 the code has not been released to PyPI [yet](https://pypi.org/project/transformers/#history) - you probably want to use `pip install git+https://github.com/huggingface/transformers.git` to get the `HEAD` that includes this PR.

I didn’t install `transformers` from PyPI, instead I installed from this repo.

younesbelkada commented:
Hi @pai4451 !
Thanks a lot for your message! This error is related to `accelerate`, I have run the colab demo this morning and everything seems to work fine. I do think that you most likely didn't installed the correct version of `accelerate` as it happened to @cnbeining before. 
Could you please share with us the output of `pip list` ? If you see that `accelerate` version is below `0.11.x` then you should re-install it with `pip install --force accelerate` 
Let us know if this works!

pai4451 commented:
> Hi @pai4451 !
> Thanks a lot for your message! This error is related to `accelerate`, I have run the colab demo this morning and everything seems to work fine. I do think that you most likely didn't installed the correct version of `accelerate` as it happened to @cnbeining before.
> Could you please share with us the output of `pip list` ? If you see that `accelerate` version is below `0.11.x` then you should re-install it with `pip install --force accelerate`
> Let us know if this works!

@younesbelkada The error is really just related to `accelerate`. After upgrading `accelerate` to `0.12` the issue is solved. Thanks for providing such a wonderful feature.

younesbelkada commented:
@pai4451 
No problem at all! I am very happy that you made it run! Let us know if you face into any other issue.

pai4451 commented:
Hi @younesbelkada, thank you again for `bitandbytes` integration with transformers models. I wonder if it is possible to use a similar way for `DeepSpeed` on int8 quantization with the BLOOM model for inference just as `bitandbytes` to transformers? Or is there any chance of loading the model using `bitandbytes` with DeepSpeed? DeepSpeed has advantages in terms of model loading and inference speed. Do you have any thoughts on how I can achieve that? I appreciate any comments you can provide.

Currently, I'm trying on the following [DeepSpeed inference script](https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/main/scripts/inference/bloom-ds-inference.py). But DeepSpeed load the model with `deepspeed.init_inference` instead of the `from_pretrained` method in transformers.

stas00 commented:
deepspeed-inference+int8 is being worked on, please give us a bit of time.

As you discovered the ds-inference script, it'll be shortly updated to support int8.

pohunghuang-nctu commented:
> deepspeed-inference+int8 is being worked on, please give us a bit of time.
> 
> As you discovered the ds-inference script, it'll be shortly updated to support int8.

Looking forward to try it, and maybe "illegal memory access" issue could be resolved by the way because of GPU memory consumption reduction on each.
